compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -6957,13 +6957,30 @@ private BType checkObjectFieldAccessExpr(BLangFieldBasedAccess fieldAccessExpr,
 
     private BType checkRecordFieldAccessExpr(BLangFieldBasedAccess fieldAccessExpr, BType varRefType, Name fieldName) {
         if (varRefType.tag == TypeTags.RECORD) {
+            BSymbol fieldSymbol = symResolver.resolveStructField(fieldAccessExpr.pos, this.env,
+                    fieldName, varRefType.tsymbol);
+
+            if (Symbols.isOptional(fieldSymbol) && !fieldSymbol.type.isNullable() && !fieldAccessExpr.isLValue) {
+                fieldAccessExpr.symbol = fieldSymbol;
+                return addNilForNillableAccessType(fieldSymbol.type);
+            }
             return checkRecordRequiredFieldAccess(fieldAccessExpr, fieldName, (BRecordType) varRefType);
         }
 
         // If the type is not a record, it needs to be a union of records.
         // Resultant field type is calculated here.
         Set<BType> memberTypes = ((BUnionType) varRefType).getMemberTypes();
 
+        // checks whether if the field symbol type is nilable and the field is optional in other records
+        for (BType memType : memberTypes) {
+            BSymbol fieldSymbol = symResolver.resolveStructField(fieldAccessExpr.pos, this.env,
+                    fieldName, memType.tsymbol);
+            if (fieldSymbol.type.isNullable() &&
+                    isFieldOptionalInRecords(((BUnionType) varRefType), fieldName, fieldAccessExpr)) {
+                return symTable.semanticError;
+            }
+        }
```

Comment:
Instead, we can move this into a method and iterate over the member type just once and maintain a count for when the field is optional (optionalCount) or is of a nilable type (nilableTypeCount).

At the end if optionalCount > 0 and nilableTypeCount > 0 we can return semanticError.